angr/analyses/decompiler/structured_codegen/c.py

```diff
@@ -1018,118 +1042,86 @@ def _c_repr_variable(self, v):
+            else:
+                self_size = 1
+            if isinstance(self.offset, int) and self.offset % self_size == 0:
+                offset = CConstant(self.offset // self_size, None, codegen=self.codegen)
+            else:
+                # uhhhhh divide by stuff
+                if isinstance(self.offset, int):
+                    offset = CConstant(self.offset, None, codegen=self.codegen)
+                else:
+                    offset = self.offset
+                offset = CBinaryOp('Div', offset, CConstant(self_size, None, codegen=self.codegen), None,
+                                   codegen=self.codegen)
 
-        elif isinstance(v_type, SimTypeArray):
-            yield from self._c_repr_array_access_with_offset(v, v_type, self_type, offset)
-            return
+            if isinstance(v_type, SimTypePointer) and v_type.pts_to == self.type:
+                yield from self._c_repr_variable(v)
+            else:
+                yield from CTypeCast(v_type, SimTypePointer(self.type).with_arch(arch), v,
+                                     codegen=self.codegen).c_repr_chunks()
 
-        elif isinstance(v, SimMemoryVariable) and isinstance(v.addr, int):
-            # loading from a global memory variable
-            yield from self._c_repr_variable(v)
+            yield "[", bracket
+            yield from offset.c_repr_chunks()
+            yield "]", bracket
             return
 
-        #
-        # other cases
-        #
-        bracket = CClosingObject("[")
+        # for stack and global variables, we need to get the address, cast to char*, add offset, and then cast to type*,
+        # and finally dereference it
+        if isinstance(v, SimMemoryVariable):
+            cv = CVariable(v, offset=0, codegen=self.codegen)
+        elif isinstance(v, CVariable) and isinstance(v.variable, SimMemoryVariable):
+            assert v.offset == 0
+            cv = v
+        else:
+            cv = None
+
+        if cv is not None:
+            cast_inner = CTypeCast(None, SimTypePointer(SimTypeChar().with_arch(arch)).with_arch(arch),
+                                   CUnaryOp("Reference", cv, None, codegen=self.codegen),
+                                   codegen=self.codegen)
+            added = CBinaryOp('Add', cast_inner, offset, None, codegen=self.codegen)
+
+            yield "*", None
+            cast_outer = CTypeCast(None, SimTypePointer(v_type).with_arch(arch), added,
+                                 codegen=self.codegen)
+            yield from cast_outer.c_repr_chunks()
+            return
 
-        # cast the variable to a pointer
-        yield from self._c_repr_variable(v)
-        yield "[", bracket
-        yield from CExpression._try_c_repr_chunks(self.offset)
-        yield "]", bracket
-        return
+        # for other variables, we simplify perform a bit shift and type cast
+        shifted = CBinaryOp('Shr', v, offset, None, codegen=self.codegen, tags=getattr(v, 'tags', None))
```

Comment:
This is using offset as a bit offset instead of a byte offset, which is a problem. In my testing, sometimes offset is in fact a bit offset, which is wildly concerning.

Reply:
Oops I'll fix it.